src/frequenz/quantities/marshmallow.py

```diff
@@ -0,0 +1,192 @@
+from ._temperature import Temperature
+from ._voltage import Voltage
+
+
+class _QuantityField(fields.Field):
+    """Custom field for Quantity objects supporting per-field serialization configuration."""
+
+    def __init__(self, field_type: Type[Quantity], **kwargs: Any) -> None:
+        """
+        Initialize the QuantityField.
+
+        Args:
+            field_type: The specific Quantity subclass (e.g., Percentage, Energy).
+            **kwargs: Additional keyword arguments for the base Field.
+
+        Raises:
+            TypeError: If field_type is not a subclass of Quantity.
+        """
+        super().__init__(**kwargs)
+        if not issubclass(field_type, Quantity):
+            raise TypeError("field_type must be a subclass of Quantity.")
+        self.field_type = field_type
+
+    def _serialize(
+        self, value: Quantity, attr: str | None, obj: Any, **kwargs: Any
+    ) -> Any:
+        """Serialize the Quantity object based on per-field configuration."""
+        if not isinstance(value, self.field_type):
+            raise ValidationError(f"Expected {self.field_type.__name__}.")
+
+        assert self.parent is not None
+
+        # Determine the serialization format
+        serialize_as_string = self.metadata.get(
+            "serialize_as_string",
+            self.parent.context.get("serialize_as_string_default", False),
+        )
+
+        if serialize_as_string:
+            # Use the Quantity's native string representation (includes unit)
+            return str(value)
+
+        # Serialize as float using the Quantity's base value
+        return value.base_value
+
+    def _deserialize(
+        self, value: Any, attr: str | None, data: Any, **kwargs: Any
+    ) -> Quantity:
+        """Deserialize the Quantity object from float or string."""
+        if isinstance(value, str):
+            # Use the Quantity's from_string method
+            return self.field_type.from_string(value)
+        if isinstance(value, (float, int)):
+            # Use `_new` method for creating instance from base value
+            return self.field_type._new(  # pylint: disable=protected-access
+                float(value)
+            )
+
+        raise ValidationError("Invalid input type for QuantityField.")
+
+
+_QUANTITY_SUBCLASSES = [
+    Current,
+    Energy,
+    Frequency,
+    Percentage,
+    Power,
+    Temperature,
+    Voltage,
+]
+
+
+def _create_quantity_field_class(
+    quantity_subclass: Type[Quantity],
+) -> Type[fields.Field]:
+    """Dynamically create a QuantityField subclass for a given Quantity subclass."""
+    class_name = f"{quantity_subclass.__name__}Field"
+
+    field_class: Type[fields.Field] = type(
+        class_name,
+        (_QuantityField,),
+        {
+            "__init__": lambda self, **kwargs: super(field_class, self).__init__(
+                field_type=quantity_subclass, **kwargs
+            ),
+            "__module__": __name__,
+        },
+    )
+
+    return field_class
```

Comment:
Oh, I would much rather have a LLM generating this once than generating type dynamically. For example the documentation won't be properly generated:

![image](https://github.com/user-attachments/assets/4fa86261-3ff5-4905-8130-e1bdd5d0cec9)

Reply:
I get the sentiment, but.. I am pretty sure no-one really needs to use those classes directly

Reply:
Like, if we want documentation for this, i would just refer people to the QuantityField class which in that case we should make public then. 

but again, I don't see any use-case why someone wants to access them.

Reply:
I agree exporting those classes is useless :P

Reply:
Though I guess that exported variable here can  be used for a users own Schema.TYPE_MAPPING, so there is that.